Source/Core/Core/IOS/USB/Emulated/Skylanders/SkylanderFigure.cpp

```diff
@@ -0,0 +1,449 @@
+
+    DecryptFigure(decrypted.data());
+
+    // Area with highest area counter is the newest
+    u16 area_offset = ((decrypted[0x89] + 1U) != decrypted[0x249]) ? 0x80 : 0x240;
+
+    figure_data.trophy_data.unlocked_villains = *(decrypted.data() + area_offset + 0x14);
+  }
+
+  return figure_data;
+}
+void SkylanderFigure::SetData(FigureData* figure_data)
+{
+  std::array<u8, FIGURE_SIZE> decrypted = {};
+
+  DecryptFigure(decrypted.data());
+
+  if (figure_data->normalized_type == Type::Skylander)
+  {
+    // Only update area with lowest counter
+    u16 area_offset = (decrypted[0x89] != (decrypted[0x249] + 1U)) ? 0x80 : 0x240;
+
+    memcpy(decrypted.data() + area_offset + 0x3, &figure_data->skylander_data.money, 2);
+    memcpy(decrypted.data() + area_offset + 0x5A, &figure_data->skylander_data.hero_level, 2);
+    memcpy(decrypted.data() + area_offset + 0x5, &figure_data->skylander_data.playtime, 4);
+
+    {
+      memcpy(decrypted.data() + area_offset + 0x60, &figure_data->skylander_data.last_reset.minute,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x61, &figure_data->skylander_data.last_reset.hour,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x62, &figure_data->skylander_data.last_reset.day, 1);
+      memcpy(decrypted.data() + area_offset + 0x63, &figure_data->skylander_data.last_reset.month,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x64, &figure_data->skylander_data.last_reset.year,
+             2);
+    }
+
+    {
+      memcpy(decrypted.data() + area_offset + 0x50, &figure_data->skylander_data.last_placed.minute,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x51, &figure_data->skylander_data.last_placed.hour,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x52, &figure_data->skylander_data.last_placed.day,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x53, &figure_data->skylander_data.last_placed.month,
+             1);
+      memcpy(decrypted.data() + area_offset + 0x54, &figure_data->skylander_data.last_placed.year,
+             2);
+    }
+
+    {
+      for (size_t i = 0; i < 8; ++i)
+      {
+        memcpy(decrypted.data() + area_offset + 0x20 + (i * 2),
+               &figure_data->skylander_data.nickname[i], 0x2);
+      }
+
+      for (size_t i = 0; i < 8; ++i)
+      {
+        memcpy(decrypted.data() + area_offset + 0x40 + (i * 2),
+               &figure_data->skylander_data.nickname[8 + i], 0x2);
+      }
+    }
+
+    {
+      ComputeChecksum(ChecksumType::Type3, decrypted.data() + area_offset + 0x50,
+                      decrypted.data() + area_offset + 0xA);
+      ComputeChecksum(ChecksumType::Type2, decrypted.data() + area_offset + 0x10,
+                      decrypted.data() + area_offset + 0xC);
+      if (decrypted[0x89] == decrypted[0x249])
+        decrypted[area_offset + 0x9]++;
+      else
+      {
+        decrypted[area_offset + 0x9] += 2;
+      }
```

Comment:
This (and similar blocks further down in the function) is very cryptic. I assume you just found this by reverse-engineering, but do you have any explanation for what's actually going on here, and can you put comments on it?

Also please either put curlies on both if-else or remove them from both.

Reply:
Ah. Yeah. Those are the area counters. To mark one as primary it needs to be one bigger then the other. So if they equal, probably a new figure, increment by one. Otherwise, increment by 2 to swap the primary one

Reply:
Ah, I see. In that case shouldn't you set the new value to be +1 of the previous value in case they're out of sync? So `decrypted[area_offset+9] = decrypted[inactive_area_offset+9] + 1`.

Reply:
Did this for all area counters